### ISSUE TITLE
How to build a release?

### ISSUE THREAD

**Issue by corpix** [NixOS/nixops#837]
Hello.
I am trying to duct tape this `StrictHostKeyChecking=no` thing https://github.com/NixOS/nixops/issues/696
To do this I decided to fork a nixops, make changes to code and build a release just for myself.
I have done this, with help of this page https://nixos.org/nixops/manual/#chap-hacking:

  - `nix-build release.nix -A build.x86_64-linux` in the root of nixops repository
  - located tarball in /nix/store with ls + grep
  - published it to github releases
  - copied the nixops derivation from nixpkgs
  - changed versions and tarball url in derivation https://github.com/corpix/nixops/blob/e097cac3e1b18a201c65c272110d11b7fdd96b31/default.nix#L5-L8

Now I am stuck with this "module not found" issue https://github.com/corpix/nixops/issues/1
How do you build nixops releases? This is so complicated :(

**Comment by srhb** [comment]
Here's an overlay I've written and used recently. I essentially grabbed all the stuff from the nixpkgs definition and added the release build. Note that I also added the missing libvirt python2 dependency.

```
self: super: {
  nixops-git = self.callPackage ({ lib, python2Packages, fetchFromGitHub, libxslt, docbook5_xsl, openssh }:
    python2Packages.buildPythonApplication rec {
      name = "nixops-${version}";
      version = "1.6pre-git";

      src = let
        src' = fetchFromGitHub {
          owner = "NixOS";
          repo = "nixops";
          rev = "f2d48f8175df1efb1535a58c9c1a83b6044bce35";
          sha256 = "1b0h020k6d94c8rwpdi6wcl09dy2gky38ffn8k2kgqpwpikl43nr";
        };
        release = self.callPackage "${src'}/release.nix" {};
      in
        "${release.tarball}/tarballs/nixops-1.6pre0_abcdef.tar.bz2";

      buildInputs = [ libxslt ];

      pythonPath = with python2Packages; [
        prettytable
        boto
        boto3
        hetzner
        libcloud
        libvirt
        azure-storage
        azure-mgmt-compute
        azure-mgmt-network
        azure-mgmt-resource
        azure-mgmt-storage
        adal
        # Go back to sqlite once Python 2.7.13 is released
        pysqlite
        datadog
        digital-ocean
      ];

    doCheck = false;

    postInstall = ''
      make -C doc/manual install nixops.1 docbookxsl=${docbook5_xsl}/xml/xsl/docbook \
        docdir=$out/share/doc/nixops mandir=$out/share/man

      mkdir -p $out/share/nix/nixops
      cp -av "nix/"* $out/share/nix/nixops

      # Add openssh to nixops' PATH. On some platforms, e.g. CentOS and RHEL
      # the version of openssh is causing errors when have big networks (40+)
      wrapProgram $out/bin/nixops --prefix PATH : "${openssh}/bin"
    '';

    meta = {
      homepage = https://github.com/NixOS/nixops;
      description = "NixOS cloud provisioning and deployment tool";
      maintainers = with lib.maintainers; [ eelco rob domenkozar ];
      platforms = lib.platforms.unix;
    };
  }) {};
}
```

**Comment by erictapen** [comment]
> This is so complicated

It's not supposed to be. Another option than doing overlays is patching Nixpkgs to use your code:

```diff
diff --git a/pkgs/tools/package-management/nixops/default.nix b/pkgs/tools/package-management/nixops/default.nix
index daecb8746ea..7842434844b 100644
--- a/pkgs/tools/package-management/nixops/default.nix
+++ b/pkgs/tools/package-management/nixops/default.nix
@@ -1,9 +1,11 @@
-{ callPackage, fetchurl }:
+{ callPackage, fetchFromGitHub }:

 callPackage ./generic.nix (rec {
   version = "1.5.2";
-  src = fetchurl {
-    url = "http://nixos.org/releases/nixops/nixops-${version}/nixops-${version}.tar.bz2";
-    sha256 = "00y2arc5rffvy6xmx4p6ibpjyc61k8dkiabq7ccwwjgckz1d2dpb";
+  src = fetchFromGitHub {
+    owner = "corpix";
+    repo = "nixops";
+    rev = "9d4d24340ab833ad53e3a0679f6eb123a0fbf061";
+    sha256 = "1cv9p4v12pdmamvrizwjaypczarr427cszbmm3k75hpxc52ndrra";
   };
 })

```

In my case I patched `nixos-17.09`, which didn't build, but I'm quite sure it should work with the `nixpkgs` version you use.
After that you have your patched `nixops` with `nixos-rebuild` or `nix-shell`.

**Comment by AmineChikhaoui** [comment]
So seems this is due to difference between pythonPath in master and in nixpkgs current nixops release so while overriding the pkg, pythonPath should be overridden as well. I just ran into the same issue few minutes ago and was able to fix by doing
```
nixpkgs.config.packageOverrides = pkgs: {
      nixUnstable = pkgs.nixUnstable.overrideAttrs (oldAttrs: rec {
        suffix = "pre5848_435ccc7";
        src = pkgs.fetchFromGitHub {
          owner = "NixOS";
          repo = "nix";
          rev = "435ccc798077e6291fd34fd1720c6abcf3521557";
          sha256 = "868712136b2dd55ac18c06a91c1a474a262184782be72d3511ed463163a898da";
        };  
      }); 
      nixops = pkgs.nixops.overrideAttrs (oldAttrs: rec {
          src = pkgs.fetchurl {
            url = "https://<PRIVATE_HYDRA>/build/2625909/download/2/nixops-1.6pre2475_f2d48f8.tar.bz2";
            sha256 = "0iqhxg08p6nl6yb79b6ar9nagdf9zf3ygd78j2128rg9bnmn6g6a";
          };
          pythonPath = with pkgs.python2Packages;
           [ prettytable
             boto
             boto3
             hetzner
             libcloud
             libvirt
             azure-storage
             azure-mgmt-compute
             azure-mgmt-network
             azure-mgmt-resource
             azure-mgmt-storage
             adal
             # Go back to sqlite once Python 2.7.13 is released
             pysqlite
             datadog
             digital-ocean
           ];
        }); 
    };
```
I'm also in favor of using Github as the source for code, the only thing done by the tarball job is preparing the manual so maybe that can be moved away in the buildPhase/installPhase but haven't looked in details how the packaging is done currently, might be missing something.

**Comment by domenkozar** [comment]
This is what we use: https://github.com/input-output-hk/iohk-ops/blob/develop/default.nix#L27

**Comment by dhess** [comment]
@domenkozar Does that evaluate in a Hydra? I tried basically the same thing you've got there, but in an overlay. 

https://github.com/dhess/nixpkgs-dhess/blob/77c1b591e8fbf77ff53cdf0bb0b76237021794da/overlays/nixops.nix#L10

It works great with an interactive `nix-build`, but when I try to build the package in Hydra, it complains about an empty job:

https://github.com/dhess/nixpkgs-dhess/blob/77c1b591e8fbf77ff53cdf0bb0b76237021794da/jobsets/release.nix

However, I did notice that you seem to be building it in your `jobsets/cardano.nix` job, so I'm confused why it's not working for me.
